text/3668-async-closure.md

```diff
@@ -0,0 +1,625 @@
+
+### Why do we even need `AsyncFnOnce`?
+
+We could desugar `async FnOnce() -> T` directly to `FnOnce<(), Output: Future<Output = T>>`. It seems overly complicated for an implementation detail, since users should never care what's *behind* the `AsyncFnOnce` trait bound.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+### Users might confusedly write `|| async {}` over `async || {}`
+
+Users may be confused whether to write `|| async {}` or `async || {}`. The fact that `async || {}` has extra "superpowers" with respect to lending may lead to users hitting unnecessary errors if they invert the ordering.
+
+We should be able to detect when users write `|| async {}` -- and subsequently hit borrow checker issues -- and give a useful error message to move the `async` keyword. We may also lint against `|| async {}` in code that *does* pass, since it's not as expressive.
+
+### Users might write `F: Fn() -> Fut, Fut: Future<Output = T>` over `F: async Fn() -> T`
+
+A similar problem could occur if users try to write "old style" trait bounds with two generic parameters `F: Fn() -> Fut` and `Fut: Future<Output = T>`.  For example:
+
+```rust!
+async fn for_each_city<F, Fut>(cb: F)
+where
+    F: Fn(&str) -> Fut,
+    Fut: Future<Output = ()>,
+{
+    for x in ["New York", "London", "Tokyo"] {
+        cb(x).await;
+    }
+}
+```
+
+This is problematic for two reasons:
+
+1. Although the `Fn` trait bound may be higher-ranked, the future that is returned cannot be, since we need to infer a single type parameter substitution for the `Future` bound to hold.
+2. There's no way for an async closure to be lending in this case, so the expressivity of the closure is limited.
+
+We can similarly implement a lint to detect cases where users write these two-part bounds and suggest that they instead write a single `async Fn() -> T` bound. This comes with the normal caveats of removing a type parameter from the function signature, e.g. semver incompatibility (since the type parameter may be turbofished). However, when users are designing a new API, they should always reach for `async Fn` trait bounds when they want to be generic over a closure that returns a future.
+
+### Lack of a desugaring
+
+It's not possible to directly name the future returned by calling some generic `T: async Fn()`. This means that it's not possible, for example, to convert `futures-rs`'s [`StreamExt::then` combinator](https://docs.rs/futures/0.3.30/futures/stream/trait.StreamExt.html#method.then), since the output future is referenced in the definition of [`Then`](https://docs.rs/futures-util/0.3.30/src/futures_util/stream/stream/then.rs.html#19) returned by the combinator.
+
+Fixing this is a follow-up goal that we're interested in pursuing in the near future.
```

Comment:
This concern isn't resolved yet -- I still need to flesh out an example that demonstrates how it's not possible to desugar a `StreamExt::then` combinator via async trait bounds with this RFC, at least not without a new `'unsafe` lifetime or unsafe binder type or some extension to the RTN syntax, and to detail possibilities of how we could do that in the future.

To be clear, I probably will not take a strong opinion on how the problem will be solved just yet, nor am I implying that this should block anything in this RFC from landing when that happens eventually, in the same way that static AFIT was not blocked on fixing the send bound problem or the dyn async trait problem since those problems can be seen as additive.

I think this is an incredibly difficult problem to solve, but it can be done on top of the abilities introduced in this RFC since I can see no way that it warrants a change to the way that async closures operate underneath the hood. That is to say, I'd like to see it be done in the near future but separately.

Reply:
That all sounds great. Thank you!

Reply:
Lost steam fleshing out this section. If you have additions please leave them, but otherwise I'm gonna resolve this comment.